homeassistant/components/lgthinq/__init__.py

```diff
@@ -0,0 +1,622 @@
+
+import asyncio
+from contextlib import suppress
+import logging
+from typing import Any
+
+from aiowebostv import WebOsClient, WebOsTvPairError
+import voluptuous as vol
+
+from homeassistant.components import notify as hass_notify
+from homeassistant.const import (
+    ATTR_COMMAND,
+    ATTR_DEVICE_ID,
+    ATTR_ENTITY_ID,
+    CONF_ACCESS_TOKEN,
+    CONF_CLIENT_SECRET,
+    CONF_COUNTRY,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_PORT,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import (
+    Event,
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+    callback,
+)
+from homeassistant.exceptions import ConfigEntryError, ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv, discovery
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.device_registry import (
+    DeviceEntry,
+    DeviceRegistry,
+    async_entries_for_config_entry,
+    async_get as async_get_device_registry,
+)
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    ATTR_BUTTON,
+    ATTR_CONFIG_ENTRY_ID,
+    ATTR_PAYLOAD,
+    ATTR_SOUND_OUTPUT,
+    CONF_CONNECT_CLIENT_ID,
+    CONF_ENTRY_TYPE,
+    CONF_ENTRY_TYPE_SOUNDBAR,
+    CONF_ENTRY_TYPE_THINQ,
+    CONF_ENTRY_TYPE_WEBOSTV,
+    DATA_CONFIG_ENTRY_WEBOSTV,
+    DATA_HASS_CONFIG,
+    DEFAULT_COUNTRY,
+    DOMAIN,
+    SERVICE_ATTR_DEVICE_INFO,
+    SERVICE_ATTR_RESULT,
+    SERVICE_ATTR_VALUE,
+    SERVICE_BUTTON,
+    SERVICE_COMMAND,
+    SERVICE_SELECT_SOUND_OUTPUT,
+    WEBOSTV_EXCEPTIONS,
+    ThinqConfigEntry,
+    ThinqData,
+)
+from .device import LGDevice, async_setup_lg_device
+from .mqtt import ThinQMQTT
+from .soundbar_client import config_connect
+from .thinq import ThinQ
+
+_LOGGER = logging.getLogger(__name__)
+
+##### common #####
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up from discovery or configuration.yaml."""
+    # hass.data only using for webostv's DATA_HASS_CONFIG and
+    # another data using entry.runtime_data
+    hass.data.setdefault(DOMAIN, {})
+    return await async_setup_webostv(hass, config)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
+    """Set up an entry."""
+    # init runtime_data
+    entry.runtime_data = ThinqData(
+        lge_device_map=None,
+        lge_devices=None,
+        lge_mqtt_clients=None,
+        soundbar_client=None,
+    )
+    if entry.data.get(CONF_ENTRY_TYPE) == CONF_ENTRY_TYPE_THINQ:
+        return await async_setup_entry_thinq(hass, entry)
+
+    if entry.data.get(CONF_ENTRY_TYPE) == CONF_ENTRY_TYPE_SOUNDBAR:
+        return await async_setup_entry_soundbar(hass, entry)
+
+    if entry.data.get(CONF_ENTRY_TYPE) == CONF_ENTRY_TYPE_WEBOSTV:
+        return await async_setup_entry_webostv(hass, entry)
+
```

Comment:
This is effectively putting 3 different integrations into a single integration; riddling the integration codebase with checks for each one of them.

This is not correct. Home Assistant supports multiple integrations being exposed under a single brand. Meaning we can make the LG brand and expose multiple integrations under it. This is what should be done here as well.

The pattern you are using here, is something we used in the very past and are no longer using or allowing.